src/transform/src/reduction_pushdown.rs

```diff
@@ -81,8 +90,707 @@ impl ReductionPushdown {
+    /// The unique keys of each of the old inputs. Avoid pushdowns where there
+    /// exists a unique key that is a subset of the group keys
+    old_input_keys: Vec<Vec<Vec<usize>>>,
+
+    // Other helper fields
+    /// The equivalences of the old join, with all single-input filters removed
+    equivalences: Vec<Vec<MirScalarExpr>>,
+    /// The support of aggregations that we do not try to push down, sorted by
+    /// input. Inner reduces must not cause these columns to be removed from
+    /// the join.
+    aggs_not_pushed_support: Vec<HashSet<usize>>,
+    /// The mapping of global to local columns in the old join
+    old_join_mapper: JoinInputMapper,
+    /// For each input, stores the inputs it is bound to and the (global) join
+    /// keys that bind it to the particular input. This is a `BTreeMap` so that
+    /// the resulting plan is deterministic.
+    join_graph: Vec<BTreeMap<usize, Vec<MirScalarExpr>>>,
+    /// whether aggregations were pushed down to any input
+    pushdown_succeeded: bool,
+}
+
+impl<'a> ReductionPusher<'a> {
+    /// Creates a struct for pushing down reductions. Returns `None` if the
+    /// relation is ineligible for reduction pushdown.
+    fn try_new(
+        group_key: &'a [MirScalarExpr],
+        aggregates: &[AggregateExpr],
+        inputs: &[MirRelationExpr],
+        equivalences: &'a [Vec<MirScalarExpr>],
+        implementation: &expr::JoinImplementation,
+        predicates: &[MirScalarExpr],
+    ) -> Option<Self> {
+        let input_types = inputs.iter().map(|i| i.typ()).collect::<Vec<_>>();
+        let old_join_mapper = JoinInputMapper::new_from_input_types(&input_types);
+
+        // sort aggregations by how and whether they can be pushed down
+        // TODO: attempt to rewrite multi-input aggregations in terms of a
+        // single input
+        // TODO: attempt to reduce the number of inputs that aggregations come
+        // from by rewriting aggregations on join keys in terms of the
+        // equivalent expression from the other input
+        let mut outer_aggs = Vec::new();
+        let mut single_input_aggs_to_push = vec![(Vec::new(), Vec::new()); inputs.len()];
+        let mut no_input_aggs_to_push = Vec::new();
+
+        // classify the aggregations by how many inputs each aggregation references
+        for (agg_num, agg) in aggregates.iter().enumerate() {
+            let mut expr_inputs = old_join_mapper.lookup_inputs(&agg.expr);
+            if agg.func.outer_agg().is_none() {
+                // Aggregates that cannot be partially pushed down go in the
+                // outer reduce.
+                outer_aggs.push((agg_num, agg.clone()));
+            } else if let AggregateFunc::Dummy = agg.func {
+                // Do not bother pushing dummy functions down.
+                outer_aggs.push((agg_num, agg.clone()));
+            } else if agg.distinct && agg.func.ignores_duplicates() {
+                outer_aggs.push((agg_num, agg.clone()));
```

Comment:
I don't understand this rule. Could you comment on the justification?

Reply:
Maybe should be `!agg.func.ignores_duplicates()`?

Reply:
Yes, it should be `!agg.func.ignores_duplicates()`. When I renamed the function to make it clearer, I changed the name of the function to have the opposite meaning but forgot to switch the location of the `!`.